File: 1/2021.11.16/391.Perfect Rectangle/java/solution.java
class Solution {
    public boolean isRectangleCover(int[][] rectangles) {
        int totalArea = 0;
        Set<String> set = new HashSet<>();
        int left = Integer.MAX_VALUE, down = Integer.MAX_VALUE, right = Integer.MIN_VALUE, up = Integer.MIN_VALUE;
        for (int[] rectangle : rectangles) {
            left = Math.min(left, rectangle[0]);
            down = Math.min(down, rectangle[1]);
            right = Math.max(right, rectangle[2]);
            up = Math.max(up, rectangle[3]);
            totalArea += (rectangle[2] - rectangle[0]) * (rectangle[3] - rectangle[1]);
            List<String> points = new ArrayList<>();
            points.add(rectangle[0] + "," + rectangle[1]);
            points.add(rectangle[2] + "," + rectangle[3]);
            points.add(rectangle[0] + "," + rectangle[3]);
            points.add(rectangle[2] + "," + rectangle[1]);
            for (String point : points) {
                if (set.contains(point)) {
                    set.remove(point);
                } else {
                    set.add(point);
                }
            }
        }
        int maxArea = (right - left) * (up - down);
        if (maxArea != totalArea) {
            return false;
        }
        if (set.size() != 4) {
            return false;
        }
        List<String> points = new ArrayList<>();
        points.add(left + "," + down);
        points.add(left + "," + up);
        points.add(right + "," + down);
        points.add(right + "," + up);
        for (String point : points) {
            if (!set.contains(point)) {
                return false;
            }
        }
        return true;
    }
}

File: 1/2021.11.16/391.Perfect Rectangle/python/solution.py
class Solution:
    def isRectangleCover(self, rectangles: List[List[int]]) -> bool:
        total_area = 0
        left = down = float('inf')
        right = up = float('-inf')
        s = set()
        for rectangle in rectangles:
            left = min(left, rectangle[0])
            down = min(down, rectangle[1])
            right = max(right, rectangle[2])
            up = max(up, rectangle[3])
            total_area += (rectangle[2] - rectangle[0]) * (rectangle[3] - rectangle[1])
            p1, p2, p3, p4 = (rectangle[0], rectangle[1]), (rectangle[2], rectangle[3]), (rectangle[0], rectangle[3]), (rectangle[2], rectangle[1])
            for p in (p1, p2, p3, p4):
                if p not in s:
                    s.add(p)
                else:
                    s.remove(p)
        return len(s) == 4 and (left, down) in s and (left, up) in s and (right, down) in s and (right, up) in s and total_area == (right - left) * (up - down)
